File: java/src/main/java/com/penapereira/patterns/visitor/ConcreteVisitor2.java
package com.penapereira.patterns.visitor;

import java.util.ArrayList;
import java.util.List;

/** Concatenates what each element computes. */
public final class ConcreteVisitor2 implements Visitor {

    private final List<String> parts = new ArrayList<>();

    @Override
    public void visitConcreteElementA(ConcreteElementA element) {
        parts.add(element.operationA());
    }

    @Override
    public void visitConcreteElementB(ConcreteElementB element) {
        parts.add(element.operationB());
    }

    public String result() {
        return String.join("+", parts);
    }
}

File: typescript/src/visitor/concrete-visitor2.ts
import type { ConcreteElementA } from "./concrete-element-a.ts";
import type { ConcreteElementB } from "./concrete-element-b.ts";
import type { Visitor } from "./visitor.ts";

/** Concatenates what each element computes. */
export class ConcreteVisitor2 implements Visitor {
  private readonly parts: string[] = [];

  visitConcreteElementA(element: ConcreteElementA): void {
    this.parts.push(element.operationA());
  }

  visitConcreteElementB(element: ConcreteElementB): void {
    this.parts.push(element.operationB());
  }

  result(): string {
    return this.parts.join("+");
  }
}
